### DISCUSSION TITLE
struct_concat?

### DISCUSSION THREAD

**Discussion by gotdan** [duckdb/duckdb#14836] [General]
Is there a way to combine two structs without knowing their keys ahead of time? 

For example, if I have the structs `{'a': 1}` and `{'b': 'one'}`, I would like to get an output like `{'a': 1, 'b': 'one'}`.

It looks like there's an undocumented (?!) `map_concat` function, but it won't work here given the mixed data types in the structs. 

Any ideas on a way to do this? If not, is there's a struct_concat function in the works?

**Comment by Maxxen** [comment]
This works:
```sql
CREATE OR REPLACE TABLE t1 (s1 STRUCT (key INT), s2 STRUCT(val VARCHAR));
INSERT INTO t1 VALUES ({key: 1}, {val: 'foo'});
-- Combine columns from both structs into one struct
SELECT combined FROM  (SELECT s1.*, s2.* FROM t1) as combined;
----
┌────────────────────────────────────┐
│              combined              │
│ struct("key" integer, val varchar) │
├────────────────────────────────────┤
│ {'key': 1, 'val': foo}             │
└────────────────────────────────────┘
```

Although it doesn't seem like its possible to fit it into a `MACRO`.

**Comment by gotdan** [comment]
Thanks @Maxxen! My data is imported json that has deeply nested structs inside lists. I think I can make the type of approach you proposed work by unnesting and then re-nesting everything, but I suspect that's going to turn out to be very slow. If there was a function that worked inside of a list_transform lambda, it would be much simpler :).

Does anyone know if struct_concat is on the duckdb roadmap, or if there's a way to create it in a reasonably performant macro?

**Reply by Maxxen** [reply]
It doesn't seem like it would be too hard to implement, but I can't really provide a timeline of when it would be implemented.

A question, what do you expect to happen if you were to `struct_concat` two structs that contain the same key? what if its the same key but the same type, or the same key but different types?

**Reply by gotdan** [reply]
It looks like `map_concat` handles duplicate keys by taking the value from the second map, which seems fine. Given this,  I'd expect a `struct_concat` function to take on the type of the second struct too. 

To be honest, the `map_concat` behavior was a little unexpected for me, since SQL functions often seem to error on this sort of thing. Maybe a `strict` flag parameter on the function that causes it to error on duplicate keys would be the most flexible option?

**Reply by Maxxen** [reply]
Made a PR #14853

**Reply by gotdan** [reply]
Awesome, thanks!
